sale_coupon_multi_use/models/sale_coupon.py

```diff
@@ -0,0 +1,154 @@
+        return new_delta, discount
+
+    def _adjust_discount_on_sale_order(self, order, amount_to_adjust):
+        self.ensure_one()
+        discount_product = self.program_id.discount_line_product_id
+        # Supposed to be only one such line.
+        coupon_line = order.order_line.filtered(
+            lambda r: r.product_id == discount_product
+        )
+        coupon_line.price_unit += amount_to_adjust
+
+    def _get_related_sale_order_line(self, sale_order):
+        self.ensure_one()
+        discount_product = self.program_id.discount_line_product_id
+        # Supposed to be only one such line.
+        return sale_order.order_line.filtered(
+            lambda r: r.product_id == discount_product
+        )[0]
+
+    def _prepare_consumption_line(self, amount_consumed, sale_order):
+        self.ensure_one()
+        line = self._get_related_sale_order_line(sale_order)
+        return {
+            "coupon_id": self.id,
+            "amount": amount_consumed,
+            "sale_order_line_id": line.id,
+        }
+
+    def _create_consumption_line(self, amount_consumed, sale_order):
+        vals = self._prepare_consumption_line(amount_consumed, sale_order)
+        return self.env["sale.coupon.consumption_line"].create(vals)
+
+    def _consume_line(self, discount, sale_order):
+        self.ensure_one()
+        amount_consumed = min(discount, self.discount_fixed_amount_delta)
+        consumption_line = self._create_consumption_line(amount_consumed, sale_order)
+        consumption_line._normalize_discount()
+        return consumption_line
+
+    def _handle_multi_use(self, coupon_order_data):
+        self.ensure_one()
+        new_delta, discount = self._get_compared_discount_with_delta(coupon_order_data)
+        self._consume_line(discount, coupon_order_data["order"])
+        # Indicates whether coupon was fully consumed.
+        return new_delta <= 0
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        """Extend to pick up coupon_multi_use value from program."""
+        for vals in vals_list:
+            multi_use = (
+                self.env["sale.coupon.program"]
+                .browse(vals.get("program_id"))
+                .coupon_multi_use
+            )
+            if "multi_use" not in vals:
+                vals["multi_use"] = multi_use
+        return super().create(vals_list)
+
+    def write(self, vals):
+        """Extend to manage multi_use coupons."""
+        if self._is_multi_use_triggered(vals):
+            coupon_order_data = self._context.get("coupon_order_data")
+            if coupon_order_data:
+                coupon_consumed = self._handle_multi_use(coupon_order_data)
+                if not coupon_consumed:
+                    # Makes it so coupon that coupon is not used up yet.
+                    del vals["state"]
+        return super().write(vals)
+
+
+class SaleCouponConsumptionLine(models.Model):
```

Comment:
Please put one model per file.

Reply:
Hm, I was using this guidelines: https://www.odoo.com/documentation/13.0/reference/guidelines.html#file-naming
It allows to group models, following main one, (like putting sale.order.line on sale_order.py, if there is also sale.order used - which is main model).

You explicitly split every model?

Reply:
@oerp-odoo Please read OCA conventions https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#15file-naming